File: java/src/main/java/com/microsoft/sample/aad/protectedapi/security/TokenValidator.java
package com.microsoft.sample.aad.protectedapi.security;


import com.microsoft.azure.spring.autoconfigure.aad.UserPrincipal;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.servlet.http.HttpServletRequest;
import lombok.NonNull;
import net.minidev.json.JSONArray;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.jwt.JwtDecoders;
import org.springframework.security.oauth2.jwt.JwtValidationException;
import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
import org.springframework.stereotype.Service;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

@Service
public class TokenValidator {

  private static final String CUSTOM_HEADER = "X-Secondary-Token";
  
  private final ActiveDirectoryConfig adConfig;

  @Autowired
  public TokenValidator(ActiveDirectoryConfig aadConfig) {
    this.adConfig = aadConfig;
  }

  /**
   * Performs basic validation of the authorization token (audience, expirancy dates and token
   * integrity).
   *
   * @return AuthContext object
   * @exception AccessDeniedException if the token is considered invalid
   */
  public AuthContext validateToken() throws AccessDeniedException {
    return validateTokenClaims(null, null);
  }

  /**
   * Performs basic validation of the authorization token including the existence of a role.
   *
   * @param requiredRole - a String containing a role that needs to be included in the authorization
   *     token
   * @return AuthContext object
   * @exception AccessDeniedException if the token is considered invalid
   */
  public AuthContext validateToken(@NonNull String requiredRole)
      throws AccessDeniedException {
    return validateTokenClaims(null, requiredRole);
  }

  /**
   * Performs basic validation of the authorization token and checks that therequester ID matches
   * the expected clientId.
   *
   * @param clientId - a String containing the clientId that must match the requester application
   *     clientId
   * @return AuthContext object
   * @exception AccessDeniedException if the token is considered invalid
   */
  public AuthContext validateTokenForClient(@NonNull String clientId)
      throws AccessDeniedException {
    return validateTokenClaims(clientId, null);
  }

  /**
   * Performs basic validation of the authorization token including the existence of a role and
   * checks that the requester ID matches the expected clientId.
   *
   * @param clientId - a String containing the clientId that must match the requester application
   *     clientId
   * @param requiredRole - a String containing a role that needs to be included in the authorization
   *     token
   * @return AuthContext object
   * @exception AccessDeniedException if the token is considered invalid
   */
  public AuthContext validateTokenForClient(
      @NonNull String clientId, @NonNull String requiredRole) throws AccessDeniedException {
    return validateTokenClaims(clientId, requiredRole);
  }

  private AuthContext validateTokenClaims(String clientId, String requiredRole)
      throws AccessDeniedException {

    // initialize AuthContext
    AuthContext ctx = new AuthContext();
    ctx.setRequiredRole(requiredRole);
    ctx.setRequestedClientId(clientId);

    // check authentication type is correct
    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
    if (!authentication.isAuthenticated()
        || !(authentication.getPrincipal() instanceof UserPrincipal)) {
      throw new AccessDeniedException("Missing or invalid Authorization token");
    }

    // validate audience
    // this should be done automatically by AADAppRoleStatelessAuthenticationFilter
    UserPrincipal principal = (UserPrincipal) authentication.getPrincipal();
    List<?> audiences = (List<?>) principal.getClaim("aud");
    String audience = audiences.get(0).toString();
    if (!audience.equals(adConfig.getAppIdUri())) {
      throw new AccessDeniedException("Incorrect audience on Authorization token");
    }
    ctx.setAudience(audience);

    // check roles
    // this can be done automatically by specifying a hasRole restriction
    // in the @PreAuthorize annotation
    Set<String> roles =
        Optional.of(principal)
            .map(p -> p.getClaim("roles"))
            .map(r -> (JSONArray) r)
            .map(Collection::stream)
            .orElseGet(Stream::empty)
            .map(Object::toString)
            .collect(Collectors.toSet());
    if (requiredRole != null && !roles.contains(requiredRole)) {
      throw new AccessDeniedException(
          String.format("Authorization token is missing required role %s", requiredRole));
    }
    ctx.setRoles(new ArrayList<String>(roles));

    // get appId from claims
    String appId = (String) principal.getClaim("appid");
    ctx.setRequesterId(appId);
    
    // set clientId and tenantId from the addConfig
    ctx.setTenantId(this.adConfig.getTenantId());
    ctx.setApiClientId(this.adConfig.getClientId());

    if (clientId != null && !appId.equals(clientId)) {
      // doesn't match - so this is either an invalid call or an aggregator
      // check if the additional custom header was supplied
      // get request header
      HttpServletRequest request =
          ((ServletRequestAttributes) RequestContextHolder.getRequestAttributes()).getRequest();

      String headerValue = request.getHeader(CUSTOM_HEADER);

      if (headerValue != null) {
        // header was supplied so we need to validate it
        ctx.setIsThirdPartyAccessAttempt(true);

        // decode secondary token
        NimbusJwtDecoder jwtDecoder =
            (NimbusJwtDecoder) JwtDecoders.fromOidcIssuerLocation(adConfig.getIssuer());
        Jwt secondaryToken = null;
        try {
          // decode token - this handles integrity checks and time validation
          secondaryToken = jwtDecoder.decode(headerValue);
        } catch (JwtValidationException exception) {
          throw new AccessDeniedException(
              String.format("%s is invalid: %s", CUSTOM_HEADER, exception.getMessage()));
        }

        String authorizedClientId = secondaryToken.getAudience().get(0);
        if (!authorizedClientId.equals("api://" + clientId)) {
          // supplied token authorizes access to a clientId different
          // from the clientId requested in the REST call
          throw new AccessDeniedException(
              String.format("Invalid audience claim on %s", CUSTOM_HEADER));
        }

        // check roles
        List<String> clientRoles = secondaryToken.getClaimAsStringList("roles");
        ctx.setClientRoles(clientRoles);

        Boolean hasClientRole = false;
        for (String role : clientRoles) {
          if (role.endsWith(requiredRole)) {
            hasClientRole = true;
            break;
          }
        }
        if (!hasClientRole) {
          throw new AccessDeniedException(
              String.format("%s token is missing required role %s", CUSTOM_HEADER, requiredRole));
        }
      } else {
        // no custom header supplied
        throw new AccessDeniedException(
            "Provided Authorization token does not grant " + "access to this resource");
      }
    } else {
      // appId matches clientId
      ctx.setIsThirdPartyAccessAttempt(false);
    }

    return ctx;
  }
}

File: dotnet/Security/TokenValidator.cs
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Primitives;
using System.Diagnostics;
using System.IdentityModel.Tokens.Jwt;
using System.Linq;
using Microsoft.Identity.Web.Resource;
using System;
using System.Security.Authentication;
using Microsoft.Extensions.Options;

namespace dotnet.Security
{
    public static class TokenValidator
    {
        private static readonly string CUSTOM_HEADER = "X-Secondary-Token";

        public static AuthContext DoTokenValidationsForClientId(AzureADConfig azureAdConfig, HttpContext ctx, string clientId, string requiredRole)
        {
            // initialize AuthContext
            AuthContext authContext = new AuthContext();
            authContext.RequiredRole = requiredRole;
            authContext.RequestedClientId = clientId;
            authContext.Authorized = false; // we default to rejecting

            //validate audience
            string audience = ctx.User.Claims.FirstOrDefault(c => c.Type == "aud")?.Value;
            if (String.IsNullOrEmpty(audience)) {
                authContext.ForbidReason = $"Authorization token is invalid";
                return authContext;
            }
            authContext.Audience = audience;

            var expectedAudience = azureAdConfig.AppIdUri;
            if (audience != expectedAudience)
            {
                authContext.ForbidReason = $"Audience was {audience} - expected {expectedAudience}";
                return authContext;
            }

            // check authorization token has access to this API method
            if (!ctx.User.IsInRole(requiredRole))
            {
                authContext.ForbidReason = $"Authorization token is missing required role {requiredRole}";
                return authContext;
            }

            //roles are valid but let's get a reference to them anyway
            string[] roles = ctx.User.Claims.Where(c => c.Type == "http://schemas.microsoft.com/ws/2008/06/identity/claims/role").Select(c => c.Value).ToArray();
            authContext.Roles = roles.ToList();

            // get appId from claims
            string appId = ctx.User.Claims.FirstOrDefault(c => c.Type == "appid").Value;
            authContext.RequesterId = appId;

            // set clientId and tenantId from the addConfig
            var tenantId = azureAdConfig.TenantId;
            var apiClientId = azureAdConfig.ClientId;
            authContext.TenantId = tenantId;
            authContext.ApiClientId = apiClientId;

            // check if it's the brand making the request
            if (appId != clientId)
            {
                // doesn't match - so this is either an invalid call or an aggregator
                // check if the additional custom header was supplied
                // get request header
                StringValues resourceHeaderValue = StringValues.Empty;
                var headerExists = ctx.Request.Headers.TryGetValue(CUSTOM_HEADER, out resourceHeaderValue);
                if (headerExists)
                {
                    authContext.IsThirdPartyAccessAttempt = true;
                    // grab secondary token 
                    var handler = new JwtSecurityTokenHandler();
                    JwtSecurityToken secondaryToken = null;
                    try
                    {
                        secondaryToken = handler.ReadJwtToken(resourceHeaderValue);
                    }
                    catch (Exception ex)
                    {
                        authContext.ForbidReason = $"Error decoding secondary token - {ex.Message}";
                        return authContext;
                    }

                    var targetClientId = secondaryToken.Audiences.FirstOrDefault();
                    if (targetClientId.StartsWith("api://"))
                    {
                        targetClientId = targetClientId.Remove(0, 6);
                    }

                    // let's validate if this is the right clientId
                    if (targetClientId != clientId)
                    {
                        // the secondary token is for another client
                        authContext.ForbidReason = $"Secondary Token is for {targetClientId} - expected {clientId}";
                        return authContext;
                    }

                    // verify brand roles
                    string[] clientRoles = secondaryToken.Claims.FirstOrDefault(c => c.Type == "roles").Value.Split(",");
                    authContext.ClientRoles = clientRoles.ToList();

                    bool hasClientRole = false;
                    foreach (string role in clientRoles)
                    {
                        if (role.EndsWith(requiredRole))
                        {
                            hasClientRole = true;
                            break;
                        }
                    }
                    if (!hasClientRole)
                    {
                        // the secondary token is for another client
                        authContext.ForbidReason = $"{CUSTOM_HEADER} is missing {requiredRole}";
                        return authContext;
                    }

                    //all required checks passed
                    authContext.Authorized = true;
                }
                else //no custom header supplied
                {
                    authContext.ForbidReason = $"Provided Authorization token does not grant access to this resource";
                    return authContext;
                }
            }
            else
            {
                // appId matches clientId
                authContext.IsThirdPartyAccessAttempt = false;
                authContext.Authorized = true;
            }

            return authContext;
        }
    }
}


